### PR TITLE
[colors] Add a "auto" color

### PR DESCRIPTION
**Pull request by olivierphi**
When text has `color: auto [percentage]` it will be displayed with the contrasting colour of its background

As a result we can have widgets that have `color: auto;` in their CSS, and see their text colour automatically adapted to their background:
Screenshot from 2022-06-29 14-33-56

### CODE REVIEW

File: src/textual/color.py [hunk 1]
```diff
@@ -145,6 +146,11 @@ def is_transparent(self) -> bool:
         """Check if the color is transparent, i.e. has 0 alpha."""
         return self.a == 0
 
+    @property
+    def is_auto(self) -> bool:
+        """Check if the color is "auto"."""
+        return self.r == AUTO.r and self.g == AUTO.g and self.b == AUTO.b
```
Comment (DrBenton) [inline src/textual/color.py:152]: Would probably be more elegant with `(self.r, self.g, self.b) == (AUTO.r, AUTO.g, AUTO.b)`, but since we're aiming at optimising for performance in most cases I guess creating a tuple takes a bit more time than doing this triple equality condition? :slightly_smiling_face:

File: sandbox/color_names.py [hunk 2]
```diff
@@ -35,8 +34,7 @@ def on_mount(self):
     def compose(self) -> ComposeResult:
         for color_name, color in COLOR_NAME_TO_RGB.items():
             color_placeholder = ColorDisplay(name=color_name)
-            is_dark_color = sum(color) < 400
-            color_placeholder.styles.color = "white" if is_dark_color else "black"
+            color_placeholder.styles.color = "auto 90%"
```
Comment (DrBenton) [inline sandbox/color_names.py:37]: This "is_dark_color" test is no longer required, since the new `auto` property does exactly this :blush: 

The result:
Screenshot from 2022-06-29 14-33-56

File: src/textual/color.py [hunk 3]
```diff
@@ -63,7 +63,8 @@ class Lab(NamedTuple):
 rgb{OPEN_BRACE}({DECIMAL}{COMMA}{DECIMAL}{COMMA}{DECIMAL}){CLOSE_BRACE}$|
 rgba{OPEN_BRACE}({DECIMAL}{COMMA}{DECIMAL}{COMMA}{DECIMAL}{COMMA}{DECIMAL}){CLOSE_BRACE}$|
 hsl{OPEN_BRACE}({DECIMAL}{COMMA}{PERCENT}{COMMA}{PERCENT}){CLOSE_BRACE}$|
-hsla{OPEN_BRACE}({DECIMAL}{COMMA}{PERCENT}{COMMA}{PERCENT}{COMMA}{DECIMAL}){CLOSE_BRACE}$
+hsla{OPEN_BRACE}({DECIMAL}{COMMA}{PERCENT}{COMMA}{PERCENT}{COMMA}{DECIMAL}){CLOSE_BRACE}$|
+(auto)(?:\s+({DECIMAL})%)?$
```
Comment (willmcgugan) [inline src/textual/color.py:67]: I'm not sure we need the decimal percentage here. We can already apply alpha with a color and a percentage.

We also don't want to enforce an order. i.e. "50% red"  should be the same as "red 50%"

File: src/textual/color.py (hunk 1 above)
Comment (willmcgugan) [reply src/textual/color.py:152]: `self[:3] == AUTO[:3]` is probably fastest. But I'm afraid I'm going to veto this sentinel color object.

File: tests/test_integration_color.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import pytest
```
Comment (willmcgugan) [inline tests/test_integration_color.py:1]: These tests run a lot of code that is unrelated to the PR.

I'd like to focus on unit tests, for instance:

- Test a color can be marked as auto
- Test Color.parse("auto") produces an auto color
- Test setting background and color on a Styles object, produces the expected `Styles.rich_style`
- Test parsing the CSS produces the expected rules

None of the above require a running app or async context. They are less likely to break from an unrelated API change.

**Review by darrenburns — APPROVED** [review]
Looks good to me :)

**Review by willmcgugan — CHANGES_REQUESTED** [review]
I think we need an alternative way of specifying the "auto" foreground color.

It may not be necessary to modify Color at all. When parsing, we could set a bool on the Styles object such as "auto_color". I think this would capture the requirement that only `color` can be auto.

Please try to minimize the use of integration tests right now. I would like to aim to have as much coverage as possible with unit tests.

File: tests/test_integration_color.py (hunk 4 above)
Comment (DrBenton) [reply tests/test_integration_color.py:1]: With only unit tests how do we check that the "auto" takes into account the colours and opacity levels of all the Node's ancestors?

Comment (willmcgugan) [reply tests/test_integration_color.py:1]: That is a good point. I think you can create a hierarchy without running the app. Create the App, call `app.mount`.

Comment (willmcgugan) [reply tests/test_integration_color.py:1]: Yeah, keep this integration test, since it's done. But let's add some unit test for what we can easily test. We may have to solve a few such issues a `active_app` to make unit tests more practical in the future. 

Integration tests are more likely to break with API changes. And the breakage is unlikely to tell me anything I didn't know already.

Integration tests will come in to their own when the APIs are more stable. They can catch unintended API changes rather than intentional API changes.

File: tests/css/test_parse.py [hunk 5]
```diff
@@ -907,18 +907,23 @@ def test_background(self):
...
+            ("rgb(1,255,50)", Color(1, 255, 50)),
+            ("rgb( 1, 255,50 )", Color(1, 255, 50)),
+            ("rgba( 1, 255,50,0.3 )", Color(1, 255, 50, 0.3)),
+            ("rgba( 1, 255,50, 1.3 )", Color(1, 255, 50, 1.0)),
+            ("hsl( 180, 50%, 50% )", Color(64, 191, 191)),
+            ("hsl(180,50%,50%)", Color(64, 191, 191)),
+            ("hsla(180,50%,50%,0.25)", Color(64, 191, 191, 0.25)),
+            ("hsla( 180, 50% ,50%,0.25 )", Color(64, 191, 191, 0.25)),
+            ("hsla( 180, 50% , 50% , 1.5 )", Color(64, 191, 191)),
+            ("red 50%", Color(255, 0, 0, 0.5)),
```
Comment (DrBenton) [inline tests/css/test_parse.py:922]: I just added these 2 cases, and Black did the rest :smile:

File: tests/test_integration_color.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        # "auto" on a white background --> foreground is black
+        ["auto", "white", BLACK],
+        # "auto" on a black background --> foreground is white
+        ["auto", "black", WHITE],
+        # And now with a pinch of alpha:
+        ["auto 50%", "white", WHITE.get_contrast_text(0.5)],
+        ["30% auto", "black", BLACK.get_contrast_text(0.3)],
+    ),
+)
+def test_color_auto_single_widget(
```
Comment (DrBenton) [inline tests/test_integration_color.py:22]: here is a "closer to the metal" test for the "auto" colour, that can only test that a Widget adapts its "auto" colour to its own background :-)

**Comment by willmcgugan** [comment]
More complicated maybe, but I think this a better abstraction. If you consider that an object should have one reason to exist, a Color class should only ever abstract a color (and "auto" is not a color).  "auto" to my mind is more of a way of enabling functionality on the Style object.

File: src/textual/css/_color.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
+"""
```
Comment (willmcgugan) [inline src/textual/css/_color.py:1]: We're doing some custom parsing here, which is different from the parsing done in the CSS. I'd like to avoid solving the same problem twice.

We could use `tokenize_value` in textual.css.tokenize for this -- which is what the CSS parser uses.

Here's an example:

```python
>>> from textual.css.tokenize import tokenize_value
>>> list(tokenize_value("hsl( 180, 50%, 50% ) 30%", "expression"))
[Token(name='color', value='hsl( 180, 50%, 50% )', path='expression', code='hsl( 180, 50%, 50% ) 30%', location=(0, 0)), Token(name='whitespace', value=' ', path='expression', code='hsl( 180, 50%, 50% ) 30%', location=(0, 20)), Token(name='scalar', value='30%', path='expression', code='hsl( 180, 50%, 50% ) 30%', location=(0, 21))]
```

This should give you the same tokens passed to the builder object.

How about the property uses `tokenize` and calls a method to extract a Color from the tokens. The builder could call this same method.

File: src/textual/css/_style_properties.py [hunk 8]
```diff
@@ -794,6 +795,37 @@ def __set__(self, obj: StylesBase, names: str | tuple[str] | None = None):
...
+
+        Args:
+            obj (Styles): The ``Styles`` object
+            value (bool | None): The value to set.
+        """
+        _rich_traceback_omit = True
+        if value is None:
+            obj.clear_rule("color_auto")
+        else:
+            obj.set_rule("color_auto", value)
```
Comment (willmcgugan) [inline src/textual/css/_style_properties.py:826]: I think "auto_color" reads a little better.

File: src/textual/css/_styles_builder.py [hunk 9]
```diff
@@ -570,31 +571,26 @@ def process_color(self, name: str, tokens: list[Token]) -> None:
...
-                    color = Color.parse(token.value)
-                except Exception as error:
-                    self.error(
-                        name,
-                        token,
-                        color_property_help_text(name, context="css", error=error),
-                    )
-            else:
-                self.error(name, token, color_property_help_text(name, context="css"))
+        expression = _join_tokens(tokens)
```
Comment (willmcgugan) [inline src/textual/css/_styles_builder.py:576]: We're parsing the color, combining the tokens, and re-parsing the string again.  See my suggestion re tokenize above. I think we can simplify by extracting the color from a list of tokens.

File: src/textual/css/styles.py [hunk 10]
```diff
@@ -182,6 +183,9 @@ class StylesBase(ABC):
     layout = LayoutProperty()
 
     color = ColorProperty(Color(255, 255, 255))
+    # N.B. This one is a pseudo-property, in the sense that we cannot set it via CSS via a dedicated property:
```
Comment (willmcgugan) [inline src/textual/css/styles.py:186]: We can also set it via code can't we? `self.styles.color_auto = True` ?

File: src/textual/css/_color.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+RE_ALPHA = (
+    # Color expressions themselves can contain percentage values (e.g. `hsl( 180, 50%, 50% )`), so we'll only
+    # consider a percentage expression to be an alpha value if it's at the start or the end of the color expression.
+    # (i.e. `color: red 50%` or `color: 50% red`)
+    re.compile(rf"^({DECIMAL})%"),  # starting with a percentage
+    re.compile(rf"({DECIMAL})%$"),  # ending with a percentage
+)
+
+
+def parse_color_expression(color_text: str) -> tuple[Color, bool]:
```
Comment (willmcgugan) [inline src/textual/css/_color.py:22]: As a pure function this is easy to test. Can we have some unit tests?

**Review by willmcgugan — CHANGES_REQUESTED** [review]
I like the update, but I think we need one more iteration on parsing the colors.

File: src/textual/css/_color.py (hunk 7 above)
Comment (DrBenton) [reply src/textual/css/_color.py:1]: My reasoning was that simply doing a `join` over the tokens in the StylesBuilder was much faster than tokenizing the values in the ColorProperty - and the 2 regexes used to find the percentage values likely have a fast lookup since they only search at the beginning and the end of a string.
But ok, I'll do that :slightly_smiling_face:

File: src/textual/css/_style_properties.py (hunk 8 above)
Comment (DrBenton) [reply src/textual/css/_style_properties.py:826]: it might read better, however my choice of `color_auto` was to collocate this property with its attached `color` as they are sister properties, so they are together when we debug for example:
Screenshot from 2022-07-01 09-15-30
It would also group the "auto" properties with their "main" one if we add more "_resolved at runtime 'auto' values_" like this one later one.

But no worries, I'll rename it :slightly_smiling_face:

Comment (willmcgugan) [reply src/textual/css/_style_properties.py:826]: No, you're right. We should follow the established convention. Ok, lets stick to `color_auto`

File: src/textual/css/styles.py (hunk 10 above)
Comment (DrBenton) [reply src/textual/css/styles.py:186]: We can yes, what my comment was meaning is that we cannot set it via a CSS string or file - as there is no `process_color_auto` method in the StylesBuilder. 
That's why to me it's a pseudo-property, as it can only be set via another CSS property, "color" :slightly_smiling_face:

Comment (willmcgugan) [reply src/textual/css/styles.py:186]: Gotcha

File: tests/test_color.py [hunk 12]
```diff
@@ -115,22 +119,28 @@ def test_color_parse(text, expected):
...
     assert Color.parse(input) == output
 
 
-@pytest.mark.parametrize("input,output", [
-    ("rgb(300, 300, 300)", Color(255, 255, 255)),
-    ("rgba(300, 300, 300, 300)", Color(255, 255, 255, 1.0)),
-    ("hsl(400, 200%, 250%)", Color(255, 255, 255, 1.0)),
-    ("hsla(400, 200%, 250%, 1.9)", Color(255, 255, 255, 1.0)),
-])
+@pytest.mark.parametrize(
```
Comment (DrBenton) [inline tests/test_color.py:135]: same here :black_circle:

File: tests/test_color.py [hunk 13]
```diff
@@ -141,7 +151,8 @@ def test_color_parse_hsl_negative_degrees():
 
 def test_color_parse_hsla_negative_degrees():
     assert Color.parse("hsla(-45, 50%, 50%, 0.2)") == Color.parse(
-        "hsla(315, 50%, 50%, 0.2)")
+        "hsla(315, 50%, 50%, 0.2)"
```
Comment (DrBenton) [inline tests/test_color.py:154]: ditto

**Comment by DrBenton** [comment]
@willmcgugan Feedback (hopefully) addressed! :slightly_smiling_face:  [28edacf](https://github.com/Textualize/textual/pull/596/commits/28edacfa57e9f72cf66399be19ff39a60bc58bc2)

File: src/textual/css/_color.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+        DeclarationError: if we come across an unexpected value
+    """
+    property_name = property_name.replace("-", "_")
+
+    alpha = 1.0
+    color = None
+    is_auto = False
+
+    for token in tokens:
+        if token.value == "auto":
```
Comment (willmcgugan) [inline src/textual/css/_color.py:36]: It's not enough to look at just the value. We also need to change "name" as well, which I think should be "token".  

Consider if we add quoted strings to the parser, we wouldn't want to interpret `"auto"` in the same way as `auto`

File: src/textual/css/styles.py [hunk 15]
```diff
@@ -684,7 +688,11 @@ def append_declaration(name: str, value: str) -> None:
             assert self.layout is not None
             append_declaration("layout", self.layout.name)
 
-        if has_rule("color"):
+        if get_rule("color_auto", False):
+            append_declaration(
+                "color", f"auto{f' {self.color.a * 100}%' if self.color.a < 1 else ''}"
```
Comment (willmcgugan) [inline src/textual/css/styles.py:693]: Can we do this without nested f strings? I find that hard to parse.

File: tests/test_color.py [hunk 16]
```diff
@@ -219,3 +230,24 @@ def test_rgb_lab_rgb_roundtrip():
...
+        ["red", Color(255, 0, 0), False],
+        ["rgba(255, 125, 33, 0.6)", Color(255, 125, 33, 0.6), False],
+        ["40% #405060", Color(64, 80, 96, 0.4), False],
+        ["hsl(45,25%,25%) 70%", Color(80, 72, 48, 0.7), False],
+        ["auto", Color(ANY, ANY, ANY, 1), True],
+        ["auto 25%", Color(ANY, ANY, ANY, 0.25), True],
+        ["32% auto", Color(ANY, ANY, ANY, 0.32), True],
+    ),
+)
+def test_parse_color_expression(
```
Comment (willmcgugan) [inline tests/test_color.py:247]: Can we have one or two invalid color strings tests? i.e. the case where it fails.

File: src/textual/css/_style_properties.py [hunk 17]
```diff
@@ -837,15 +870,18 @@ def __set__(self, obj: StylesBase, color: Color | str | None):
             if obj.set_rule(self.name, color):
                 obj.refresh()
         elif isinstance(color, str):
+            tokens = tokenize_value(color, "expression")
             try:
-                parsed_color = Color.parse(color)
+                parsed_color, is_auto = parse_color_tokens(self.name, tokens)
```
Comment (willmcgugan) [inline src/textual/css/_style_properties.py:875]: This can raise `DeclarationError`s can't it?

This would be entirely unexpected in this context. It should raise `StyleValueError`s

File: src/textual/css/styles.py (hunk 15 above)
Comment (willmcgugan) [inline src/textual/css/styles.py:693]:
```suggestion
                "color", f"auto {self.color.a * 100}%" if self.color.a < 1 else "auto"
```

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Consider this...

```python
self.styles.color = "red"
```

Fantastic, we have red text. I now do this:

```python
self.styles.color = "auto 90%"
```

Great, the text is now white-ish or black-ish. All good so far.

```python
self.styles.color_auto = False
````

Now what color is the text? I suspect it will be `Color(1,2,3, a=0.9)` which would be a surprise to the developer.

My gut feeling is that it should be red again.

I think we should solve this. Could you have a think about a solution and let's have a chat about it when you have a plan?

**Comment by darrenburns** [comment]
I have to say I'm really not a fan of the `color_auto` - it's so tightly coupled to `color` and having 2 properties interfering with each other like that is just confusing. If we have to guess what the behaviour is for something that is conceptually so simple, I think we're already going to the wrong route.

I liked the idea of a sentinel value but agree the previous approach broke the "color" abstraction.

IMO it'd be nice to have the color property value have a type `Color | Auto` (i.e. a union where `Auto` is some container or sentinel to mark the special case), rather than having a separate flag. The approach as it stands feels like "using a flag to simulate a union type".

**Comment by willmcgugan** [comment]
So `self.styles.color` could return a Color or an Auto?

What kind of interface would the Auto support? Can I write `self.styles.color.r` or would I have to do `is isinstance(self.styes.color)` beforehand?

**Comment by willmcgugan** [comment]
I did not mean to do that!

**Comment by willmcgugan** [comment]
I agree that `styles.color` should probably not be anything other than a Color. It would surprise me if that wasn't the case..

For the Python interface, I think we will need `color_auto` as a boolean, but also `color_auto_alpha` as a float to capture the alpha color. I think that would avoid the edge case I mention above.

@DrBenton I see where you are going with another CSS property to capture the auto color state, but find the declaration of "color: auto 90%" to be quite elegant. And another rule might confuse, especially if there is a "color" *and* a "color-auto".

**Comment by willmcgugan** [comment]
Just for the record. After a 3 way call I think we've agreed on the following:

- The `Styles.color` property *setter* can accept "auto" with an optional alpha, e.g. "auto 90%"
- The `Styles.color` property *getter* returns the calculated contrasting color.
- We add a `Styles.color_auto` boolean (*getter* only) that exposes if "auto" is in effect.

I think this alleviates any edge cases, and is reasonably easy to wrap ones head around.

Thumbs up if you're in favor...

**Comment by DrBenton** [comment]
@willmcgugan here you go! [4439ca1](https://github.com/Textualize/textual/pull/596/commits/4439ca1308926c8c5ac44e60f31ea66c385df4e4)

File: src/textual/css/_style_properties.py [hunk 18]
```diff
@@ -808,14 +810,47 @@ def __get__(
...
+        # We're dealing with a "color" property that has an "auto" value: let's traverse the Node ancestors to
+        # calculate what the resulting value should be.
+        from ..dom import DOMNode
+
+        styles_node = getattr(obj, "node", None)
+        if not isinstance(styles_node, DOMNode):
+            raise ValueError(
+                "Dynamic 'auto' color can only be used with an subclass of StylesBase that has a 'node' attribute"
+            )
+        # We just repeat what `DOMNode.color` does, in a slightly simplified way.
```
Comment (willmcgugan) [inline src/textual/css/_style_properties.py:836]: Have you considered rolling this logic in to the `color` property itself?

File: src/textual/css/_style_properties.py [hunk 19]
```diff
@@ -808,14 +810,47 @@ def __get__(
...
+
         Args:
             obj (Styles): The ``Styles`` object.
             objtype (type[Styles]): The ``Styles`` class.
 
         Returns:
             Color: The Color
         """
-        return cast(Color, obj.get_rule(self.name, self._default_color))
+        if self.name != "color" or not obj.color_auto:
```
Comment (willmcgugan) [inline src/textual/css/_style_properties.py:823]: Rather than overload the functionality of this object based on its name. We could define a property specifically for the foreground color.

This would make it easier to keep the functionality distinct. 

I think this may already be an issue. For instance, if you do `styles.background = "auto 90%"` I'm guessing that would set the foreground to auto? Which would not be desirable.

File: src/textual/css/_style_properties.py (hunk 18 above)
Comment (DrBenton) [reply src/textual/css/_style_properties.py:836]: The thing is that `DOMNode.colors` returns 4 computed colours - foreground and background colours, and its parent's colours. Whereas the Style properties each manage their single property

File: src/textual/css/_style_properties.py [hunk 20]
```diff
@@ -891,13 +860,74 @@ def __set__(self, obj: StylesBase, color: Color | str | None):
...
+        styles_node = getattr(obj, "node", None)
+        if not isinstance(styles_node, DOMNode):
+            raise ValueError(
+                "Dynamic 'auto' color can only be used with an subclass of StylesBase that has a 'node' attribute"
+            )
+        # We just repeat what `DOMNode.color` does, in a slightly simplified way.
+        # N.B. `DOMNode.color` will call this `__get__` method while it is itself traversing the DOMNode ancestors,
+        # so we loop over ancestors in a loop that is itself looping over ancestors.
+        # That's a deliberate implementation choice.
+        background = Color(0, 0, 0, 0)
```
Comment (DrBenton) [inline src/textual/css/_style_properties.py:913]: this dynamic resolution is now only made for the "color" property

**Comment by DrBenton** [comment]
@willmcgugan I added a new ForegroundColorProperty, and moved the dynamic resolution of "_color: auto_" there: [ea1a429](https://github.com/Textualize/textual/pull/596/commits/ea1a429c412dcfa2e014f9653a93b5f7d28eceb4)

Compared to the very first one, this implementation is:
 - 3-4 times more verbose (in terms of lines of code)
 - More complex
 - Slower (the runtime cost of the 1st implementation was just a one-line "if" added in the DomNode's `colors` getter)
 - And all this without any difference in the end result, or the public API that Textual users will see... So, well, I must say that I'm not a big fan of it myself :sweat_smile: 

But if it fits better your idea of how the design for that new feature should look like, I guess that's worth the cost :slightly_smiling_face:
